### PR TITLE
issue126 allow arbitrary users

### PR DESCRIPTION
**Pull request by makuk66**
Use the user mechanisms from https://docs.openshift.org/latest/creating_images/guidelines.html#use-uid

Locally this now works:
```
mkdir mycores mylogs
docker run -it -e VERBOSE=yes \
  -e USER_NAME=$(id -un) \
  --user $(id -u) \
  -p 8983:8983 \
  -v $PWD/mycores:/opt/solr/server/solr/mycores\
  -v $PWD/mylogs:/opt/solr/server/logs \
  docker-solr/docker-solr:6.6.0-openshift \
  solr-precreate core1
```

For #126

### CODE REVIEW

File: 5.5/alpine/scripts/add-user [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    set -x
+fi
+
+if [[ ! -z "$NO_ADD_USER" ]]; then
+  exit 0
+fi
+
+if ! whoami &> /dev/null; then
+  if [ -w /etc/passwd ]; then
+    echo "${USER_NAME:-default}:x:$(id -u):0:${USER_NAME:-default} user:${HOME}:/sbin/nologin" >> /etc/passwd
```
[the remark this answers is not in the document]
Comment (makuk66) [reply 5.5/alpine/scripts/add-user]: The `solr` user is created in the Dockerfile, and this arbitrary user is created at runtime. The only way for it to be `solr` would be to not create a user in the Dockerfile, default to store files as root and always run as root, and then always create the solr user and chown it at runtime; I don't think I like that...

Why do you want it to be `solr` -- aesthetics or does something else depend on it?

**Comment by tianon** [comment]
The ability to run the image as an arbitrary user is definitely something we love to see (see docker-library/rabbitmq#60, docker-library/cassandra#48, docker-library/mongo#81, docker-library/redis#48, docker-library/mysql#161, docker-library/mariadb#59, docker-library/percona#21, and docker-library/ghost#54, https://github.com/docker-library/postgres/pull/253 for some good examples where we've done exactly that), but we almost never do so via an environment variable given that Docker itself already has `--user` which enforces the UID switch before the container ever starts, and when combined with `--security-opt no-new-privileges` can provide really strong constraints on the container process.

I'd also honestly recommend very strongly against anything that touches `/etc/passwd` at runtime.  If users need a specific UID and Solr doesn't support running as a completely arbitrary UID, then I'd recommend two things.  For Solr, I'd recommend fixing upstream to not require that so strongly (since it's probably not an intentional limitation -- using something like `id -u` returning `0` or non-`0` will be much more reliable for testing whether or not we're `root` than `whoami`, for example).  For users, I'd recommend using user namespaces, making a new image `FROM` this one which adds the specific UID they're interested in using to `/etc/passwd`, or bind-mounting `/etc/passwd` from their host (which likely has the UID they're interested in using already defined, and is one of the main reasons I'd recommend against ever trying to modify `/etc/passwd` at container run-time).

**Comment by jamiejackson** [comment]
I just realized that the solution discussion was happening over here, versus #126. @tianon, could you take a look at @GrahamDumpleton's [reply](https://github.com/docker-solr/docker-solr/issues/126#issuecomment-353490427), which includes a defense of `/etc/passwd` updates?

**Comment by GrahamDumpleton** [comment]
FWIW, Red Hat recommending making ``/etc/passwd`` writable in this context of running in a container was only after it went through many many levels of security review by lots of Linux security gurus so everyone was happy that it doesn't result in any security issues.

**Comment by mhutter** [comment]
What's the status here @makuk66 ? It's still not possible to run solr on OpenShift, something we'd love to do 👍

**Comment by jamiejackson** [comment]
@GrahamDumpleton, would you happen to have a reference to the "many many levels of security review by lots of Linux security gurus" that gives credence your approach in the [patch you provided](https://github.com/docker-solr/docker-solr/issues/126#issuecomment-353490427)? I feel your solution is the most considered (with consensus from a greater community), but I fear that it is being dismissed/ignored and that Solr's implementation will be insular and short-sighted. (Sorry, Solr developers, I think you're great! but I sense things veering off track in this particular area.)

**Comment by GrahamDumpleton** [comment]
No public reference really, was internal people at Red Hat.

BTW, since then have found that there is one extra thing that is also needed though when using the approach of making ``passwd`` file writable. You need to configure PAM to only allow people in group ``wheel`` to do ``su``. This is necessary where image is used in container runtimes where Linux capability for doing ``su`` is NOT dropped. Except for OpenShift, most platforms don't drop the capability, so you must have it if image is not OpenShift specific.

So in ``Dockerfile`` also do:

```
echo "auth required pam_wheel.so use_uid" >> /etc/pam.d/su
```

Example can be seen at:

* https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile#L48

Discussion on issue at:

* https://github.com/jupyter/docker-stacks/issues/560#issuecomment-384114575

Would also ensure get latest ``fix-permissions`` script, as may have been updated, from Red Hat repo at:

* https://github.com/sclorg/s2i-base-container/blob/master/core/root/usr/bin/fix-permissions

FWIW, I had it on my TODO list to write an updated blog post on the topic of how to set image up to run as non ``root``. Was hoping to do it after finishing current blog post series am working on.

**Comment by makuk66** [comment]
Thanks for those references Graham, I'll have to read them in some more detail. From a once-over, it seems that the maintainer for that project shares my concerns about the root group chown approach with regards to security risk, compatibility for other environments, and efficiency. For docker-solr, with it being part of the docker library, I'm also wary of taking an approach that is not shared by other packages there.

Jamie wrote:

> and that Solr's implementation will be insular and short-sighted.

First we figure out the story with upstream Solr 8 and the other requirements, then we see if/how we can accomodate special environments; be it by default, or with runtime configuration, or through sub images. Once there is a PR we can discuss specifics.

**Comment by makuk66** [comment]
To summarise my thoughts on this subject, at this time:

- OpenShift's implentation choices mean images need to have some non-standard properties.
  That creates an onus on them to provide such images, and indeed they do so in their Catalog and https://github.com/sclorg/.
  So with regards to having "custom images just for OpenShift", really, it is up to them to provide that and I think your requests should be directed there.
  Now, docker-solr can make such support easier, at various levels: by making it easy to create sub images (which is where we are today), or provide tagged images on Docker hub (which would be convenient for users), or in the same tags if that does not negatively affect other users
  and does not break too much with tradition in the official docker library (and that I'm less conviced is the case).
- I dislike the root group file ownership plus the running with the root group. For normal use it's just not required, it's not like other images in the official docker library, and it does mark a change from what we do at the moment, and could thus affect existing users. That's what I'm not keen to see in our default tags.
  I'm not ruling it out or in for `solr:8` at this point; but once `solr:8` happens, that will be fixed and I'm not going to discuss it endlessly.
- I really dislike /etc/passwd modification and PAM mods. Fundamentally Solr itself shouldn't care.
  The use of "whoami" in Solr's scripts is something that should be fixed upstream, for which I have proposes a patch upstream [SOLR-13087](https://issues.apache.org/jira/browse/SOLR-13087) and it would be helpful if people commented/upvoted there. In the meantime, docker-solr already patches that.
The ugly `I have no name!@1f6fe1697d11:/opt/solr$` prompt is cosmetic, and I can live with (and I have patched too).
So I don't see a valid reason we'd have to actually have a user in /etc/passwd to correspond to the OpenShift user id; please correct me with concrete examples if I'm wrong.
- I'm not going to put in the docker library an incompatible intermediate `solr:7.x` version, which I then have to explain/document/maintain, which may change in months for Solr 8. You'll just have to wait a little longer.
- I do however want to have an exprimental version we can use to verify what works in OpenShift (and elsewhere), as soon as possible.
- If need be, we could have a `solr:8-openshift` variant of the images.
- I'm not advocating running as root. Solr refuses to start as root, so that wouldn't work anyway. There are some use-cases for starting as root and then dropping to solr with `gosu`, and we support that today, in as far as you being able to construct a bash invocation (or mount some script) to do so.

So, here's what I've done:

- Not specific to OpenShift, but I've created a new branch https://github.com/docker-solr/docker-solr/pull/202 . This moves the bash logic from the Dockerfile RUN statements to scripts, which then get COPYd and RUN.
This makes it easier to check with shellcheck, easier to maintain, easier to extend.
Tradeoffs of this approach: fewer layers, and thus slower `docker build` in cases where scripts are modified (solr needs re-downloading); this doesn't affect normal use. And there are a few wasted K in a layer. I think it's worth it. I wanted this with multi-stage builds, but we can't do that yet.
I intend to merge that.
- I've created a new `openshift` branch off that, with a separate openshift variant. That makes things root group writable at docker build time. There's a PR https://github.com/docker-solr/docker-solr/pull/203. I'm happy to tweak that further for openshift if needed
- I've pushed an initial image of that to `dockersolr/docker-solr:7.6.0-openshift`. Please experiment with that, but do not rely on it, as it may change, and will disappear once `solr:8` happens.
  (note that the other `dockersolr/docker-solr` tags track `master`, not this branch).
  As always, you can build this locally and experiment.
- With this, the following works: `docker run --user 7777:0 docker-solr/docker-solr:7.6.0-openshift solr-create -c gettingstarted`.
- I've tried that image on OpenShift, and it appears to work, without any additional config required.
- I've made some changes to the new Dockerfile (replace some ENVs with ARGs) which I want to do for `solr:8`, so can try here. As that's a change to the environment that current users might rely on (though I doubt it), I think I'll do this only for 8.

What needs to happen next:
 
- I'm closing this PR; discuss the new PR there.
- Please test out `docker-solr/docker-solr:7.6.0-openshift` in OpenShift, and report what doesn't; then we can iterate there.
- Test out `docker-solr/docker-solr:7.6.0-openshift` in non-OpenShift (mac, linux, dc/os, k8s; with volumes), and report what doesn't work (or report that it does); hopefully nothing.
- Test out what scenarios are not addressed. Between normal, gosu, and openshift, presumably none.
- I'll continue to work with the Solr 8 devs to see what's next for directory layout (this is expected to be a matter of weeks)
- I've done exploratory work for the init_solr_service layout change for https://github.com/docker-solr/docker-solr/pull/204. That looks promising
- Based on the outcome of all that, we can debate wether we merge the `-openshift` tag as a separate variant, or backport its script to other variants. @tianon and the Docker library folks can help guide there.

**Comment by yosifkit** [comment]
Regarding the moving of setup from the Dockerfile into scripts:

> It may be tempting, for the sake of brevity, to put complicated initialization details into a standalone script and merely add a RUN command in the Dockerfile. However, this causes the resulting Dockerfile to be overly opaque, and such Dockerfiles are unlikely to pass review. Instead, it is recommended to put all the commands for initialization into the Dockerfile as appropriate RUN or ENV command combinations. 
>
> \- https://github.com/docker-library/official-images/tree/47d50af6e2cd2128b4f1c2ddcc1228d8d7724db6#clarity

-------------

I thought their was a discussion somewhere, but I can only find one [really long comment](https://github.com/docker-library/official-images/pull/517#issuecomment-76439816) about images that are custom for a provider/platform, so let me summarize my thoughts (I'll try to work this into the official-images readme soon):

Adding extra variants to images (or adding new images) that are only for a specific vendor platform (Amazon, Google Cloud, OpenShift, OpenStack, etc) breaks the whole "run anywhere" docker experience. Images in Official Images should be platform independent. It is up to image maintainers whether or not to adjust their images to better support the specific needs of a platform in a general way.

**Comment by makuk66** [comment]
Thanks for your comments @yosifkit.

> It may be tempting, for the sake of brevity, to put complicated initialization details into a standalone script 

I understand the sentiment behind the desire to keep things in the Dockerfile, and for simple scenarios I'd agree. However, there comes a point where the amount of logic becomes too large. In particular when:
- you start incorporating control constructs like [nested loops](https://github.com/docker-solr/docker-solr/blob/external-and-updates/7.6/scripts/build-tools/get-gpg-keys.sh#L10)
- you have [long commands](https://github.com/docker-solr/docker-solr/blob/external-and-updates/7.6/scripts/build-tools/get-solr.sh#L35) which will need further logic to do the right thing for different versions.
- you start re-using common logic in different parts of your Dockerfiles, like [the gpg key fetching](https://github.com/docker-solr/docker-solr/blob/external-and-updates/7.6/Dockerfile#L35)
- code gets replicated to multiple different variant Dockerfile templates by hand, creating opportunity for mismatches which are not easy to see.
- between the templates and the generated Dockerfiles (27 in total) you end up with large diffs during review, which is overly opaque
- the code becomes large and complex enough that you really want the ability to run static analysis tools like spellcheck for prevention of bugs.

I see that not as some "for the sake of brevity" convenience, but as decomposition, DRY, quality control and good engineering practice.

Those are my arguments, let me know if you consider them compelling. If you still say it will not pass review, I'll change it begrudingly.

> I can only find one really long comment about images that are custom for a provider/platform

That thread refers to a somewhat different scenario: vmtp appears to be a tool specific to one cloud environment, not a general purpose application like Solr, and I agree that that may not be a good fit. 

> Images in Official Images should be platform independent. It is up to image maintainers whether or not to adjust their images to better support the specific needs of a platform in a general way.

Allright, so out of my options (create a variant in the official images, add root group ownership in all the official images, create a separate repo/variant/image/whatever outside the official images, or just say no), the first one is out. Fair enough. 

Are you aware of any other images in the official images doing root group ownership to achieve compatibility on OpenShift?

**Pull request closed without merging**